### ISSUE TITLE
This is not a issue, but thoughts about view filtering

### ISSUE THREAD

**Issue by Ciantic** [philix/fxql#1]
I was exploring to create a similar tooling but you seem to have started this already. I've read about GraphQL and it would be natural to think it as a filtering language over a existing view to database.

Take for instance this:

```
node(701763886) {
  id,
  name,
  birthdate {
    month,
    day
  },
  friends.after(73014066062280).first(2) {
    cursor,
    node {
      name
    }
  }
}
```

It's actually just a filter over a view that returns the user. Now how do we make a user view to database that returns everything in user joined with Slick and then take the GraphQL as a filter for that view? I have no good answers yet, but the basic idea is you have REST like actions e.g.:

```scala
 def user = Action.async(parse.graphQL) { implicit request =>
  // Do query that returns single user with everything joined, given with the argument `701763886`
  // Filter the query resuls with GraphQL and return only the given fields
}
```

Above is inefficient way to do GraphQL filtering, but it works, gets the ball rolling. This way one could build the system gradually, first implement a views that return everything in the table joined, and then filter the response naively before response. 

"Final" step is to revise the filtering algorithm and turn it to full fledged slick filters:

```scala
 def user = Action.async(parse.graphQL) { implicit request =>
  // Create query that returns single user with everything joined, and filter using Slick methods from the GraphQL
}
```

This is a harder step which would make the GraphQL efficient and the end goal, but the point is first we could figure out how to make it naively filter result set from slick.

**Comment by philix** [comment]
By naively filtering you mean, fetching everything and filtering manually for every view (user, comments...)?

I built the parser with the intuition that we can automatically map GraphQL queries to Slick which implements the harder functional to relational mapping.

**Comment by Ciantic** [comment]
Yes that is what I meant. I do recognize that on list of things like "comments" this probably does not work very well, the result set is too big. I have not really read any of your code yet, though it would be nice to have real thing. Is the FXQL a subset of GraphQL or does it include some 'extensions'?

I gather this is pretty early, but I'm exploring for a strict subset because Facebook is working on static analysis tools on the client side of things, they probably wouldn't work with fork of the GraphQL. Some form of lite-GraphQL server that implements even the field name splicing would be useful at the beginning.

**Comment by philix** [comment]
@Ciantic my code implements only the parser (i.e. it builds a abstract syntax tree of the query) and it was done based on the first presentation about the language.

Since they recently released the spec (http://facebook.github.io/graphql/) we can start implementing the real thing.

Notice that even if they release a GraphQL implementation it won't be generally useful. You still have to figure out how to translate the query (even if it's already parsed) to queries to your data store. I think that mapping GraphQL queries to Slick queries would be great.

You can already start playing with GraphQL implementation in Ruby and JavaScript (https://medium.com/@clayallsopp/your-first-graphql-server-3c766ab4f0a2).
